File: primatis-api/src/main/java/be/primatis/setting/web/SettingResponse.java
package be.primatis.setting.web;

import be.primatis.setting.ApplicationSetting;
import be.primatis.user.AppUser;

import java.time.Instant;
import java.util.Objects;

/**
 * Contrat REST de lecture d'un {@code application_setting} (DEV-12.2,
 * {@code GET}/{@code PATCH /api/v1/settings}). Expose exactement les champs
 * attendus par le mandat DEV-12.2 §6 (DTO de lecture) : {@code settingKey},
 * {@code settingValue}, {@code valueType}, {@code description}, {@code
 * updatedAt}, {@code updatedByUser}. Ne reproduit jamais l'Entity {@code
 * ApplicationSetting} telle quelle.
 *
 * <p>{@code updatedByUser} est {@code null} tant qu'aucune modification
 * n'a encore eu lieu depuis le bootstrap Flyway (V001/V006, {@code
 * updated_by_user_id} absent de l'INSERT initial) — reflète strictement
 * l'état persisté, jamais une valeur de repli substituée.
 */
public record SettingResponse(
        String settingKey,
        String settingValue,
        String valueType,
        String description,
        Instant updatedAt,
        SettingUpdatedByResponse updatedByUser) {

    public static SettingResponse from(ApplicationSetting setting) {
        Objects.requireNonNull(setting, "setting");
        AppUser updatedByUser = setting.getUpdatedByUser();
        return new SettingResponse(
                setting.getSettingKey(),
                setting.getSettingValue(),
                setting.getValueType(),
                setting.getDescription(),
                setting.getUpdatedAt(),
                updatedByUser == null ? null : SettingUpdatedByResponse.from(updatedByUser));
    }
}

File: primatis-web/src/app/settings/models/setting-response.ts
import { SettingUpdatedByResponse } from './setting-updated-by-response';
import { SettingValueType } from './setting-value-type';

/**
 * Voir `be.primatis.setting.web.SettingResponse` côté backend
 * (`GET`/`PATCH /api/v1/settings`, DEV-12.2). Dates ISO conservées en
 * chaîne, jamais converties en `Date` ici (frontend.md « Dates »).
 *
 * `updatedAt: string | null` reprend le typage explicitement demandé par le
 * mandat DEV-12.3 §8. En pratique, le backend ne renvoie jamais `null` pour
 * ce champ (`ApplicationSetting.updatedAt` est `NOT NULL`, bootstrappé par
 * Flyway V001/V006, DEV-12.1 §8) — seul `updatedByUser` est réellement
 * nullable tant qu'aucune modification administrative n'a eu lieu. Le
 * typage nullable est conservé tel quel par prudence (aucun contrat
 * runtime validé, comme tout DTO de ce projet) ; la page traite
 * `updatedByUser === null` comme le signal réel de « jamais modifié »
 * (DÉDUIT MÉCANIQUEMENT, cf. log DEV-12.3 §5).
 */
export interface SettingResponse {
  readonly settingKey: string;
  readonly settingValue: string;
  readonly valueType: SettingValueType;
  readonly description: string;
  readonly updatedAt: string | null;
  readonly updatedByUser: SettingUpdatedByResponse | null;
}
